commit bf6ac302cde4aef55ded7e5340f16e6cd8b8e874
Author: Maciej W. Rozycki <macro@redhat.com>
Date:   Fri Mar 28 12:35:52 2025 +0000

    stdio-common: Add tests for formatted sscanf input specifiers
    
    Wire sscanf into test infrastructure for formatted scanf input
    specifiers.
    
    Reviewed-by: Joseph Myers <josmyers@redhat.com>

diff --git a/stdio-common/Makefile b/stdio-common/Makefile
index e8e06f194e6592b6..2892f680c069836a 100644
--- a/stdio-common/Makefile
+++ b/stdio-common/Makefile
@@ -52,7 +52,7 @@ nonfmt-xprintf-stems := \
 xprintf-stems := $(sort $(fmt-xprintf-stems) $(nonfmt-xprintf-stems))
 
 # List of markers for scanf family function tests.
-xscanf-funcs := s f
+xscanf-funcs := s f ss
 
 # List of data types and formats for individual per-conversion scanf tests.
 # Further conversions are provided by sysdeps.
diff --git a/stdio-common/tst-scanf-format-ss-c.c b/stdio-common/tst-scanf-format-ss-c.c
new file mode 100644
index 0000000000000000..58a7ce8f63398f22
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-c.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for the character conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-character.h"
+#include "tst-scanf-format-skeleton-c.c"
diff --git a/stdio-common/tst-scanf-format-ss-char.c b/stdio-common/tst-scanf-format-ss-char.c
new file mode 100644
index 0000000000000000..66eb04c0d620126d
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-char.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for signed char conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-integer.h"
+#include "tst-scanf-format-skeleton-char.c"
diff --git a/stdio-common/tst-scanf-format-ss-double.c b/stdio-common/tst-scanf-format-ss-double.c
new file mode 100644
index 0000000000000000..2a9fd8afb552fa96
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-double.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for double conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-real.h"
+#include "tst-scanf-format-skeleton-double.c"
diff --git a/stdio-common/tst-scanf-format-ss-float.c b/stdio-common/tst-scanf-format-ss-float.c
new file mode 100644
index 0000000000000000..640e35dbebd2e459
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-float.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for float conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-real.h"
+#include "tst-scanf-format-skeleton-float.c"
diff --git a/stdio-common/tst-scanf-format-ss-int.c b/stdio-common/tst-scanf-format-ss-int.c
new file mode 100644
index 0000000000000000..c01a0a3f1f1a530d
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-int.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for int conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-integer.h"
+#include "tst-scanf-format-skeleton-int.c"
diff --git a/stdio-common/tst-scanf-format-ss-ldouble.c b/stdio-common/tst-scanf-format-ss-ldouble.c
new file mode 100644
index 0000000000000000..3e57ab568f45a741
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-ldouble.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for long double conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-real.h"
+#include "tst-scanf-format-skeleton-ldouble.c"
diff --git a/stdio-common/tst-scanf-format-ss-llong.c b/stdio-common/tst-scanf-format-ss-llong.c
new file mode 100644
index 0000000000000000..9c1603c869b7f5dc
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-llong.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for long long conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-integer.h"
+#include "tst-scanf-format-skeleton-llong.c"
diff --git a/stdio-common/tst-scanf-format-ss-long.c b/stdio-common/tst-scanf-format-ss-long.c
new file mode 100644
index 0000000000000000..003cfac55282b3cc
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-long.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for long conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-integer.h"
+#include "tst-scanf-format-skeleton-long.c"
diff --git a/stdio-common/tst-scanf-format-ss-short.c b/stdio-common/tst-scanf-format-ss-short.c
new file mode 100644
index 0000000000000000..ba1a58289cd7bd1d
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-short.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for short int conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-integer.h"
+#include "tst-scanf-format-skeleton-short.c"
diff --git a/stdio-common/tst-scanf-format-ss-uchar.c b/stdio-common/tst-scanf-format-ss-uchar.c
new file mode 100644
index 0000000000000000..1db14241dcab6d89
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-uchar.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for unsigned char conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-integer.h"
+#include "tst-scanf-format-skeleton-uchar.c"
diff --git a/stdio-common/tst-scanf-format-ss-uint.c b/stdio-common/tst-scanf-format-ss-uint.c
new file mode 100644
index 0000000000000000..a296fb48c9c9500a
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-uint.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for unsigned int conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-integer.h"
+#include "tst-scanf-format-skeleton-uint.c"
diff --git a/stdio-common/tst-scanf-format-ss-ullong.c b/stdio-common/tst-scanf-format-ss-ullong.c
new file mode 100644
index 0000000000000000..885d5709be386471
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-ullong.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for unsigned long long int conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-integer.h"
+#include "tst-scanf-format-skeleton-ullong.c"
diff --git a/stdio-common/tst-scanf-format-ss-ulong.c b/stdio-common/tst-scanf-format-ss-ulong.c
new file mode 100644
index 0000000000000000..8a2f24dc5ce7d8d8
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-ulong.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for unsigned long int conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-integer.h"
+#include "tst-scanf-format-skeleton-ulong.c"
diff --git a/stdio-common/tst-scanf-format-ss-ushort.c b/stdio-common/tst-scanf-format-ss-ushort.c
new file mode 100644
index 0000000000000000..40618889ac34e910
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss-ushort.c
@@ -0,0 +1,22 @@
+/* Test for formatted 'sscanf' input for unsigned short int conversions.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "tst-scanf-format-skeleton.h"
+#include "tst-scanf-format-ss.h"
+#include "tst-scanf-format-integer.h"
+#include "tst-scanf-format-skeleton-ushort.c"
diff --git a/stdio-common/tst-scanf-format-ss.h b/stdio-common/tst-scanf-format-ss.h
new file mode 100644
index 0000000000000000..2fb1ca20deb845df
--- /dev/null
+++ b/stdio-common/tst-scanf-format-ss.h
@@ -0,0 +1,73 @@
+/* Test feature wrapper for formatted 'scanf' input.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <support/support.h>
+
+static char *sscanf_buf;
+static size_t sscanf_buf_size;
+
+static void __attribute__ ((destructor))
+scanf_under_test_fini (void)
+{
+  free (sscanf_buf);
+}
+
+#define scanf_under_test(...)						\
+({									\
+  __label__ out;							\
+  size_t i = 0;								\
+  int result;								\
+  int ch;								\
+									\
+  do									\
+    {									\
+      ch = read_input ();						\
+      if (ch < 0)							\
+	{								\
+	  result = ch;							\
+	  goto out;							\
+	}								\
+      if (i == sscanf_buf_size)						\
+	{								\
+	  sscanf_buf_size += SIZE_CHUNK;				\
+	  /* Add an extra byte for the terminating null character.  */	\
+	  sscanf_buf = xrealloc (sscanf_buf, sscanf_buf_size + 1);	\
+	}								\
+      sscanf_buf[i++] = ch;						\
+    }									\
+  while (ch != ':');							\
+  sscanf_buf[i++] = '\0';						\
+									\
+  ch = ungetc (ch, stdin);						\
+  if (ch == EOF)							\
+    {									\
+      result = INPUT_ERROR;						\
+      goto out;								\
+    }									\
+									\
+  result = sscanf (sscanf_buf, __VA_ARGS__);				\
+  if (result == EOF)							\
+    result = INPUT_EOF;							\
+									\
+out:									\
+  result;								\
+})
